xwebrtc/workspace.go

```diff
@@ -0,0 +1,236 @@
+package xwebrtc
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"cdr.dev/slog"
+	"github.com/pion/webrtc/v3"
+	"golang.org/x/xerrors"
+	"nhooyr.io/websocket"
+
+	"cdr.dev/coder-cli/coder-sdk"
+	"cdr.dev/coder-cli/pkg/proto"
+)
+
+const (
+	// NetworkTCP is the protocol for tcp tunnels.
+	NetworkTCP = "tcp"
+)
+
+// WorkspaceClient dials workspace agents and represents peer connections as a http.Client.
+type WorkspaceClient struct {
+	log           slog.Logger
+	brokerAddr    *url.URL
+	token         string
+	httpClient    *http.Client
+	wsDialOptions *websocket.DialOptions
+}
+
+// NewWorkspaceClient creates a new workspace client to dial to agents.
+func NewWorkspaceClient(log slog.Logger, brokerAddr *url.URL, token string) *WorkspaceClient {
+	client := &WorkspaceClient{
+		log:        log,
+		brokerAddr: brokerAddr,
+		token:      token,
+	}
+	client.httpClient = &http.Client{
+		Transport: &http.Transport{
+			DialContext: client.DialContext,
+		},
+	}
+	client.wsDialOptions = &websocket.DialOptions{
+		HTTPClient: client.httpClient,
+	}
+
+	return client
+}
+
+// DialContext will create a new peer connection with the workspace agent, make a new data channel, and return it as
+// a net.Conn.
+func (wc *WorkspaceClient) DialContext(ctx context.Context, network string, workspaceAddr string) (net.Conn, error) {
+	workspaceID, port, err := ParseWorkspaceAddr(workspaceAddr)
+	if err != nil {
+		return nil, xerrors.Errorf("parsing workspace addr: %w", err)
+	}
+
+	wc.log.Debug(ctx, "parsed wsp addr", slog.F("id", workspaceID), slog.F("port", port))
+	rtc, err := wc.PeerConnection(ctx, workspaceID)
```

Comment:
I'd rather if we made the `WorkspaceClient` create the connection on `NewWorkspaceClient` for a specific workspace, then our `DialContext` arguments can be normal instead of a customized format.

Reply:
Updated to make the active peer connection in the workspace client and renamed to WorkspaceDialer